File: RDMPEG/RDMPEGDecoder/RDMPEGSubtitleASSParser/RDMPEGSubtitleASSParser.m
//
//  RDMPEGSubtitleASSParser.m
//  RDMPEG
//
//  Created by Serhii Alpieiev on 10/3/17.
//  Copyright © 2017 Readdle. All rights reserved.
//

#import "RDMPEGSubtitleASSParser.h"



NS_ASSUME_NONNULL_BEGIN

@implementation RDMPEGSubtitleASSParser

+ (nullable NSArray<NSString *> *)parseEvents:(NSString *)events {
    NSRange range = [events rangeOfString:@"[Events]"];
    if (range.location != NSNotFound) {
        NSUInteger position = range.location + range.length;
        
        range = [events rangeOfString:@"Format:"
                          options:0
                            range:NSMakeRange(position, events.length - position)];
        
        if (range.location != NSNotFound) {
            position = range.location + range.length;
            range = [events rangeOfCharacterFromSet:[NSCharacterSet newlineCharacterSet]
                                        options:0
                                          range:NSMakeRange(position, events.length - position)];
            
            if (range.location != NSNotFound) {
                NSString *format = [events substringWithRange:NSMakeRange(position, range.location - position)];
                NSArray *fields = [format componentsSeparatedByString:@","];
                
                if (fields.count > 0) {
                    NSMutableArray *events = [NSMutableArray array];
                    for (NSString *field in fields) {
                        [events addObject:[field stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]]];
                    }
                    return events;
                }
            }
        }
    }
    
    return nil;
}

+ (nullable NSArray<NSString *> *)parseDialogue:(NSString *)dialogue numFields:(NSUInteger)numFields {
    if ([dialogue hasPrefix:@"Dialogue:"]) {
        NSMutableArray *fields = [NSMutableArray array];
        
        NSRange range = NSMakeRange(@"Dialogue:".length, 0);
        NSUInteger currentField = 0;
        
        while (range.location != NSNotFound && currentField < numFields) {
            const NSUInteger position = range.location + range.length;
            
            range = [dialogue rangeOfString:@","
                                    options:0
                                      range:NSMakeRange(position, dialogue.length - position)];
            
            const NSUInteger length = (range.location == NSNotFound || currentField == numFields - 1) ? (dialogue.length - position) : (range.location - position);
            
            NSString *field = [dialogue substringWithRange:NSMakeRange(position, length)];
            field = [field stringByReplacingOccurrencesOfString:@"\\N" withString:@"\n"];
            [fields addObject:field];
            
            currentField++;
        }
        
        return fields;
    }
    
    return nil;
}

+ (NSString *)removeCommandsFromEventText:(NSString *)text {
    NSMutableString *result = [NSMutableString string];
    
    NSScanner *scanner = [NSScanner scannerWithString:text];
    while (scanner.isAtEnd == NO) {
        NSString *s;
        if ([scanner scanUpToString:@"{\\" intoString:&s]) {
            [result appendString:s];
        }
        
        if (!([scanner scanString:@"{\\" intoString:nil] &&
              [scanner scanUpToString:@"}" intoString:nil] &&
              [scanner scanString:@"}" intoString:nil])) {
            break;
        }
    }
    
    return result;
}

@end

NS_ASSUME_NONNULL_END

File: RDMPEG/RDMPEGDecoder/RDMPEGSubtitleASSParser/RDMPEGSubtitleASSParser.swift
//
//  RDMPEGSubtitleASSParser.swift
//  RDMPEG
//
//  Created by Max Berezhnoy on 15/07/2024.
//  Copyright © 2024 Readdle. All rights reserved.
//

import Foundation

@objcMembers
public class RDMPEGSubtitleASSParser: NSObject {

    public class func parseEvents(_ events: String) -> [String]? {
        guard let range = events.range(of: "[Events]") else { return nil }
        var position = range.upperBound

        guard let formatRange = events.range(of: "Format:", range: position..<events.endIndex) else { return nil }
        position = formatRange.upperBound

        guard let newlineRange = events.rangeOfCharacter(
            from: .newlines,
            range: position..<events.endIndex
        ) else {
            return nil
        }

        let format = events[position..<newlineRange.lowerBound]
        let fields = format.components(separatedBy: ",")

        guard fields.isEmpty == false else { return nil }

        return fields.map { $0.trimmingCharacters(in: .whitespaces) }
    }

    public class func parseDialogue(_ dialogue: String, numFields: UInt) -> [String]? {
        guard dialogue.hasPrefix("Dialogue:") else { return nil }

        var fields: [String] = []
        var range = dialogue.index(dialogue.startIndex, offsetBy: "Dialogue:".count)..<dialogue.endIndex
        var currentField: UInt = 0

        while range.lowerBound != dialogue.endIndex && currentField < numFields {
            let position = range.lowerBound

            if let commaRange = dialogue.range(of: ",", range: position..<dialogue.endIndex) {
                range = commaRange.upperBound..<dialogue.endIndex
            }
            else {
                range = dialogue.endIndex..<dialogue.endIndex
            }

            let field = dialogue[position..<range.lowerBound]
                .replacingOccurrences(of: "\\N", with: "\n")
            fields.append(String(field))

            currentField += 1
        }

        return fields
    }

    public class func removeCommandsFromEventText(_ text: String) -> String {
        var result = ""
        let scanner = Scanner(string: text)
        scanner.charactersToBeSkipped = nil

        while scanner.isAtEnd == false {
            if let scanned = scanner.scanUpToString("{\\") {
                result += scanned
            }

            if scanner.scanString("{\\") != nil,
               scanner.scanUpToString("}") != nil,
               scanner.scanString("}") != nil {
                continue
            }
            else {
                break
            }
        }

        return result
    }
}
